middleware/proxy.go

```diff
@@ -287,19 +302,52 @@ func ProxyWithConfig(config ProxyConfig) echo.MiddlewareFunc {
 				req.Header.Set(echo.HeaderXForwardedFor, c.RealIP())
 			}
 
-			// Proxy
-			switch {
-			case c.IsWebSocket():
-				proxyRaw(tgt, c).ServeHTTP(res, req)
-			case req.Header.Get(echo.HeaderAccept) == "text/event-stream":
-			default:
-				proxyHTTP(tgt, c, config).ServeHTTP(res, req)
-			}
-			if e, ok := c.Get("_error").(error); ok {
-				err = e
-			}
+			retries := config.RetryCount
+			for {
+				var tgt *ProxyTarget
+				var err error
+				if isTargetProvider {
+					tgt, err = provider.NextTarget(c)
+					if err != nil {
+						return err
+					}
+				} else {
+					tgt = config.Balancer.Next(c)
+				}
+				c.Set(config.ContextKey, tgt)
+
+				// Proxy
+				switch {
+				case c.IsWebSocket():
+					proxyRaw(tgt, c).ServeHTTP(res, req)
+				case req.Header.Get(echo.HeaderAccept) == "text/event-stream":
+				default:
+					proxyHTTP(tgt, c, config).ServeHTTP(res, req)
+				}
 
-			return
+				e, hasError := c.Get("_error").(error)
```

Comment:
I would change this block to:

```go
				err, hasError := c.Get("_error").(error)
				if !hasError {
					return nil
				}

				if rErr := config.ErrorHandler(err, c); rErr != nil {
					return rErr
				}

				if retries <= 0 {
					return err
				}

				retries--
			}
```

I looked into `proxyHTTP` function and from that I think that checking `httpErr.Code == http.StatusBadGateway` is making too much assumptions in behalf of the developer.  Default behavior to  stop on every error is aligned how `proxyHTTP` works. It can return 2 variants `echo.NewHTTPError(StatusCodeContextCanceled` or `echo.NewHTTPError(http.StatusBadGateway` and all these cases are perfectly reasonable for not retrying anymore.

I removed `c.Set(config.ContextKey, nil)` because this will be always overridden anyway. Also not clearing target here allows developer to check which target fails when `tgt, err = provider.NextTarget(c)` is called. as at that time that target would be still in context.

and `c.Set("_error", nil)` was moved up so `provider.NextTarget(c)` could check it and after that we clear error from context.

Reply:
edit: I made assumption that we can return on every error. After good night of sleep I understand that this would not play ok with retry. so I edited default errorhandler to discard error when it sees `echo.NewHTTPError(http.StatusBadGateway`.  This is almost same as your variant but allows developer full control over errors.

Reply:
I am happy to make this change if you insist, but I also feel that it's very, very unlikely the other errors that could be returned would be resolved by retrying the request, or a user would ever want to do that. 

Looking through the code, they mostly seem to be things that would be considered "fatal" errors in another langauge. We can provide the flexibility, but I think that comes at a cost of making it more complex for users to provide their own retry filter. For example, if I want to only retry GET requests, in my custom function I now need to also make sure I am checking for the same `http.StatusBadGateway` AND then add my own check after that. Also for a user to provide a custom error check, they would need to go through the internal Proxy code and find specific errors they can check for, since these are not documented or described clearly (which makes sense as most are likely fatal, internal-type errors that are unrecoverable anyway). I also feel the feature is about retrying proxy requests when a backend server is down, not retrying requests from internal fatal errors.

I feel like if we make this change, we are making the interface more complex for almost every user, on the very small chance that some users might want to retry errors that probably won't be fixed by retrying anyway.

What do you think?